File: sdks/canary/typescript/scenarios/s-ws-status.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// S-WS-STATUS canary — TypeScript SDK

import { LLMSafeSpaces } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch, rawDo, hasField } from '../canary.js';

async function run(r: Runner, cfg: Config): Promise<void> {
  const c = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: cfg.apiKey, timeout: 20000, fetch: nodeFetch as any });
  let wsId: string | null = null;
  try {
    const [ok, ws] = await r.assertNoError(
      () => c.workspaces.create({ name: 'canary-ts-wsstatus', runtime: 'base', storageSize: '1Gi' }),
      'create: no error');
    if (!ok || !ws) return;
    wsId = ws.id;

    const [ok2, st] = await r.assertNoError(() => c.workspaces.getStatus(wsId!), 'get-status: no error');
    if (ok2 && st) {
      // Phase may be empty pre-reconcile (canary CI runs a controller-less
      // cluster) — same relaxation as the Go/Python twins.
      r.assert('phase' in st && typeof st.phase === 'string', 'status: phase present (may be empty pre-reconcile)');
      r.assert(typeof st.activeSessions === 'number' && st.activeSessions >= 0, 'status: activeSessions ≥ 0');
      r.assert('credentialState' in st, 'status: credentialState present');
      r.assert('agentHealth' in st, 'status: agentHealth present');
      r.assert(typeof (st as any).agentHealth?.status === 'string', 'status: agentHealth.status is string');
    }

    // Raw — no error field on success
    const [s, b] = await rawDo('GET', `${cfg.apiUrl}/api/v1/workspaces/${wsId}/status`, cfg.apiKey);
    r.assert(s === 200, `status-raw: 200 (got ${s})`);
    r.assert(!hasField(b, 'error'), 'status-raw: no error field');

  } finally {
    if (wsId) { try { await c.workspaces.delete(wsId); } catch {} }
  }

  await r.assertError(() => c.workspaces.getStatus('00000000-0000-0000-0000-000000000000'),
    'status-nonexistent: error');
}

async function main() {
  const r = new Runner('ws-status');
  await run(r, configFromEnv());
  r.print();
  process.exit(r.exitCode());
}
main().catch(e => { console.error(e); process.exit(1); });

File: sdks/canary/python/scenarios/s_ws_status.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""S-WS-STATUS canary — Python SDK"""

from __future__ import annotations
import sys, os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
from canary import Runner, Config, config_from_env, raw_do, has_field
from llmsafespaces import LLMSafeSpaces, NotFoundError


def run(r: Runner, cfg: Config) -> None:
    c = LLMSafeSpaces(cfg.api_url, api_key=cfg.api_key, timeout=20.0)
    ws_id = None
    try:
        ok, ws = r.assert_no_error(
            lambda: c.workspaces.create(
                name="canary-py-wsstatus", runtime="base", storage_size="1Gi"
            ),
            "create: no error",
        )
        if not ok:
            return
        ws_id = ws.id

        ok2, st = r.assert_no_error(
            lambda: c.workspaces.get_status(ws_id), "get-status: no error"
        )
        if ok2 and st is not None:
            # Phase may be empty immediately after creation — the
            # controller hasn't reconciled yet (Pending → Creating →
            # Active). In canary CI there may be no controller running,
            # so accept empty phase for a freshly-created workspace; the
            # contract under test is that GetStatus returns 200 with a
            # parseable shape (mirror of the Go twin).
            r.assert_(
                st.get("phase", "") is not None,
                "status: phase present (may be empty pre-reconcile)",
            )
            r.assert_(
                isinstance(st.get("activeSessions"), int) and st["activeSessions"] >= 0,
                "status: activeSessions ≥ 0",
            )
            r.assert_("credentialState" in st, "status: credentialState present")
            r.assert_("agentHealth" in st, "status: agentHealth present")
            r.assert_(
                isinstance(st.get("agentHealth", {}).get("status"), str),
                "status: agentHealth.status is string",
            )
            # Conditions may be empty pre-Active — the field must be
            # parseable (present-or-absent without crash), not non-empty
            # (mirror of the Go twin).
            r.assert_(
                st.get("conditions") is None or isinstance(st.get("conditions"), list),
                "status: conditions field parseable",
            )

        # Raw — no error field on success
        s, b = raw_do(
            "GET", f"{cfg.api_url}/api/v1/workspaces/{ws_id}/status", cfg.api_key
        )
        r.assert_(s == 200, f"status-raw: 200 (got {s})")
        r.assert_(not has_field(b, "error"), "status-raw: no error field")

    finally:
        if ws_id:
            try:
                c.workspaces.delete(ws_id)
            except:
                pass

    # N1: nonexistent
    r.assert_error(
        lambda: c.workspaces.get_status("00000000-0000-0000-0000-000000000000"),
        "status-nonexistent: error",
    )


if __name__ == "__main__":
    r = Runner("ws-status")
    cfg = config_from_env()
    run(r, cfg)
    r.print()
    sys.exit(r.exit_code())
